### PR TITLE
chore: use `resolve-package-path` to resolve installed dependency versions more reliably

### LINKED ISSUE #25782: Launchpad CT onboarding dependencies screen not always reliable
### Current behavior

When doing the onboarding for CT, sometimes it claims you are missing a dependency, when you are not. Eg in this example I have both Solid.js and Vite Plugin Solid, but it says I don't:

image

Problem is here: https://github.com/cypress-io/cypress/blob/develop/packages/scaffold-config/src/frameworks.ts#L26-L28

Error:

```
cypress:scaffold-config:frameworks error when detecting solid-js: Package subpath './package.json' is not defined by "exports" in /Users/lachlanmiller/code/dump/my-solid-app/node_modules/solid-js/package.json 
```

While it is possible to fix, eg [like this](https://github.com/solidjs/solid/pull/1538/files#diff-24f04605b9a786a23c3cf2ab9649fb66cf01126ed9c7a202100840b6a8323fa1R250), we need to have an alternative strategy that does not rely on `package.json` getting added to the export map for every module on npm.

### Desired behavior

Correctly find and determine the module and version.

### Test code to reproduce

The easiest way to reproduce would be

```
yarn create vite my-app --template react
```

`cd` into `node_modules/react`, remove `package.json` from the `exports` key in `package.json`.

Do the Cypress setup and observe it does not detect that you have React installed.

### Cypress Version

12

### Node version

16

### Operating System

Linux, macos

### Debug Logs

```shell
cypress:scaffold-config:frameworks error when detecting solid-js: Package subpath './package.json' is not defined by "exports" in /Users/lachlanmiller/code/dump/my-solid-app/node_modules/solid-js/package.json 
```
```


### Other

_No response_

### PR DESCRIPTION
**Pull request by astone123**
- Closes #25782

### Additional details

There are some dependencies that we can't resolve with [`require.resolve`](https://nodejs.org/api/modules.html#requireresolverequest-options), because of one of the following reasons:
- The package has `exports` defined, but `package.json` is not one of the exports
- Some packages don't have entrypoints (like in the case of [`react-scripts`](https://github.com/facebook/create-react-app/tree/main/packages/react-scripts), there is no entrypoint, just a bin).

We need to find a way around these limitations so that we can figure out if a dependency is installed, and if it is, which version is installed. Lachlan and I found a package called [`resolve-package-path`](https://www.npmjs.com/package/resolve-package-path) that does the trick, working around the `exports` and entrypoint limitations and allowing us to detect dependencies that are installed.

TLDR: It looks like we can't use `require.resolve` to figure out which dependencies of which packages are installed.

[Further reading](https://stackoverflow.com/questions/10111163/how-can-i-get-the-path-of-a-module-i-have-loaded-via-require-that-is-not-mine)

### Steps to test

1. `yarn create vite my-app --template react`
2. Install dependencies in `my-app`
3. cd into `node_modules/react`, remove `package.json` from the exports key in `package.json`
4. Open Launchpad with `my-app` and click Component Testing
5. Verify that the framework and bundler (React and Vite) are automatically detected and are selected in the dropdowns

### How has the user experience changed?

There should be no noticeable change for the user

### PR Tasks

- [x] Have tests been added/updated?
- [na] Has a PR for user-facing changes been opened in [`cypress-documentation`](https://github.com/cypress-io/cypress-documentation)? 
- [na] Have API changes been updated in the [`type definitions`](https://github.com/cypress-io/cypress/blob/develop/cli/types/cypress.d.ts)?

### CODE REVIEW

File: packages/launchpad/cypress/e2e/scaffold-component-testing.cy.ts [hunk 1]
```diff
@@ -83,6 +83,26 @@ describe('scaffolding component testing', {
       cy.findByRole('button', { name: 'Continue' }).click()
       verifyConfigFile(`cypress.config.ts`)
     })
+
+    it('detects react dependency even if `package.json` is not declared in `exports`', () => {
+      cy.scaffoldProject('react-vite-ts-unconfigured')
```
Comment (lmiller1990) [inline packages/launchpad/cypress/e2e/scaffold-component-testing.cy.ts:88]: We have a create-react-app fixture, should we just use that? Then you won't need to hack into the `package.json` - it already lacks `main` and `exports`.

File: packages/scaffold-config/src/frameworks.ts [hunk 2]
```diff
@@ -38,7 +42,7 @@ export async function isDependencyInstalled (dependency: Cypress.CypressComponen
     return {
       dependency,
       detectedVersion: pkg.version,
-      loc,
+      loc: path.parse(packageFilePath).dir,
```
Comment (lmiller1990) [inline packages/scaffold-config/src/frameworks.ts]: I am curious if this works for namespaced packages like `@vitejs/plugin-react` which also lacks `package.json` in `exports`: https://github.com/vitejs/vite-plugin-react/blob/main/packages/plugin-react/package.json

Can you verify that works as expected? Not sure how `dir` will resolve.

Also - do we even use `.loc`? If not, we could just remove it entirely.

File: packages/launchpad/cypress/e2e/scaffold-component-testing.cy.ts (hunk 1 above)
Comment (astone123) [reply packages/launchpad/cypress/e2e/scaffold-component-testing.cy.ts:88]: Right, but I wanted to test against a project that has `exports` defined, but doesn't export `package.json` since that's where we run into the issue

File: packages/scaffold-config/src/frameworks.ts (hunk 2 above)
Comment (astone123) [reply packages/scaffold-config/src/frameworks.ts]: Yeah I don't see this being used anywhere anyway. I'll remove it

File: packages/launchpad/cypress/e2e/scaffold-component-testing.cy.ts (hunk 1 above)
Comment (lmiller1990) [reply packages/launchpad/cypress/e2e/scaffold-component-testing.cy.ts:88]: Right, I see, sure thing.

File: packages/scaffold-config/src/frameworks.ts [hunk 3]
```diff
@@ -17,11 +17,14 @@ export type CodeGenFramework = Cypress.ResolvedComponentFrameworkDefinition['cod
 export async function isDependencyInstalled (dependency: Cypress.CypressComponentDependency, projectPath: string): Promise<Cypress.DependencyToInstall> {
   try {
     debug('detecting %s in %s', dependency.package, projectPath)
-    const loc = require.resolve(path.join(dependency.package, 'package.json'), {
-      paths: [projectPath],
-    })
 
-    const pkg = await fs.readJson(loc) as PkgJson
+    const packageFilePath = resolvePackagePath(dependency.package, projectPath) || ''
```
Comment (lmiller1990) [inline packages/scaffold-config/src/frameworks.ts]: Does this also work as expected for nested modules like `@vitejs/plugin-react` etc?

Comment (astone123) [reply packages/scaffold-config/src/frameworks.ts]: yeah, I tested it with a different project
```
> resolvePackagePath('@vitejs/plugin-react', resolve('.'))
'/Users/adams/projects/vite-react-17/node_modules/@vitejs/plugin-react/package.json'
```

File: system-tests/projects/react-vite-ts-unconfigured/package.json [hunk 4]
```diff
@@ -8,12 +8,12 @@
...
-    "react-dom": "^17.0.2"
+    "react": "^18.0.0",
+    "react-dom": "^18.0.0"
   },
   "devDependencies": {
-    "@types/react": "^17.0.33",
-    "@types/react-dom": "^17.0.10",
+    "@types/react": "^18.0.0",
+    "@types/react-dom": "^18.0.0",
     "@vitejs/plugin-react": "^1.0.7",
```
Comment (mike-plummer) [inline system-tests/projects/react-vite-ts-unconfigured/package.json]: Pretty old version of Vite & the React plugin - worth updating?

File: packages/scaffold-config/src/frameworks.ts (hunk 3 above)
Comment (mike-plummer) [inline packages/scaffold-config/src/frameworks.ts]: Maybe I'm missing something, but couldn't we do this a bit more simply?

```suggestion
    const packageFilePath = resolvePackagePath(dependency.package, projectPath)
    if (!packageFilePath) {
      throw Error('Unable to resolve dependency')
    }
```

File: packages/scaffold-config/src/frameworks.ts [hunk 5]
```diff
@@ -38,7 +41,6 @@ export async function isDependencyInstalled (dependency: Cypress.CypressComponen
     return {
       dependency,
       detectedVersion: pkg.version,
-      loc,
       satisfied,
     }
   } catch (e) {
```
Comment (mike-plummer) [inline packages/scaffold-config/src/frameworks.ts:52]: Total nitpick, but for this debug logging is a failure to lookup a dependency really an "error"? We have the ability to differentiate an error (filesystem permissions, etc) from a "dependency is missing" condition. Could report each differently by doing an early return above rather than throwing an Error to be caught here

File: packages/scaffold-config/package.json [hunk 6]
```diff
@@ -18,6 +18,7 @@
     "debug": "^4.3.4",
     "fs-extra": "^9.1.0",
     "globby": "^11.0.1",
+    "resolve-package-path": "4.0.3",
```
Comment (mike-plummer) [inline packages/scaffold-config/package.json]: Any reason this is pinned to `4.0.3` rather than doing a `^4.0.3`?

File: system-tests/projects/react-vite-ts-unconfigured/package.json (hunk 4 above)
Comment (astone123) [reply system-tests/projects/react-vite-ts-unconfigured/package.json]: sure, why not

File: packages/scaffold-config/src/frameworks.ts (hunk 3 above)
Comment (astone123) [reply packages/scaffold-config/src/frameworks.ts]: yeah good call

File: packages/scaffold-config/src/frameworks.ts (hunk 5 above)
Comment (astone123) [reply packages/scaffold-config/src/frameworks.ts:52]: Good point, I updated this so that it just logs to debug that we couldn't resolve the dependency - no need to throw an error

File: packages/scaffold-config/package.json (hunk 6 above)
Comment (astone123) [reply packages/scaffold-config/package.json]: No reason, added it manually and forgot to add `^`

**Comment by mike-plummer** [comment]
@astone123 Still seeing a lot of test failures here, happy to ✅ once those are resolved

**Comment by astone123** [comment]
@mike-plummer okay we should be good now

**Review by mike-plummer — APPROVED** [review]

**Review by lmiller1990 — APPROVED** [review]
